src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1269,5 +1226,270 @@ internal override bool CanAddProjectReference(ProjectId referencingProject, Proj
             var result = referenceManager.QueryCanReferenceProject(referencingHierarchy, referencedHierarchy);
             return result != (uint)__VSREFERENCEQUERYRESULT.REFERENCE_DENY;
         }
+
+        private readonly object _gate = new object();
```

Comment:
>private readonly object _gate = new object(); [](start = 8, length = 45)

Move to top.

Reply:
Going to leave these here for now -- in some sense we may want to move this over to a separate partial...